evidence/pool.go

```diff
@@ -177,35 +283,162 @@ func (evpool *Pool) MarkEvidenceAsCommitted(height int64, evidence []types.Evide
 
 		if err := evpool.evidenceStore.Set(key, evBytes); err != nil {
 			evpool.logger.Error("Unable to add committed evidence", "err", err)
-			// if we can't move evidence to committed then don't remove the evidence from pending
-			continue
-		}
-		// if pending, remove from that bucket, remember not all evidence has been seen before
-		if evpool.IsPending(ev) {
-			evpool.removePendingEvidence(ev)
-			blockEvidenceMap[evMapKey(ev)] = struct{}{}
 		}
 	}
 
 	// remove committed evidence from the clist
 	if len(blockEvidenceMap) != 0 {
 		evpool.removeEvidenceFromList(blockEvidenceMap)
 	}
+
+	return abciEvidence
 }
 
-// Has checks whether the evidence exists either pending or already committed
-func (evpool *Pool) Has(evidence types.Evidence) bool {
-	return evpool.IsPending(evidence) || evpool.IsCommitted(evidence)
+// EvidenceFront goes to the first evidence in the clist
+func (evpool *Pool) EvidenceFront() *clist.CElement {
+	return evpool.evidenceList.Front()
 }
 
-// IsEvidenceExpired checks whether evidence is past the maximum age where it can be used
-func (evpool *Pool) IsEvidenceExpired(evidence types.Evidence) bool {
-	return evpool.IsExpired(evidence.Height(), evidence.Time())
+// EvidenceWaitChan is a channel that closes once the first evidence in the list is there. i.e Front is not nil
+func (evpool *Pool) EvidenceWaitChan() <-chan struct{} {
+	return evpool.evidenceList.WaitChan()
+}
+
+// SetLogger sets the Logger.
+func (evpool *Pool) SetLogger(l log.Logger) {
+	evpool.logger = l
+}
+
+// State returns the current state of the evpool.
+func (evpool *Pool) State() sm.State {
+	evpool.mtx.Lock()
+	defer evpool.mtx.Unlock()
+	return evpool.state
+}
+
+//--------------------------------------------------------------------------
+
+// Info is a wrapper around the evidence that the evidence pool receives with extensive
+// information of what validators were malicious, the time of the attack and the total voting power
+// This is saved as a form of cache so that the evidence pool can easily produce the ABCI Evidence
+// needed to be sent to the application.
+type Info struct {
+	Evidence         types.Evidence
+	Time             time.Time
+	Validators       []*types.Validator
```

Comment:
a bit excessive that we store whole valset (could be hundreds of vals) per evidence

Reply:
Okay so my thinking is that evidence will naturally be sparse and so memory concerns aren't too important. The advantage of saving them like this is that we don't have to retrieve the entire validator set and then work out which of the validators were malicious thus it makes it easier to form abci Evidence

Reply:
Also in the case of duplicate vote evidence it's just a single validator per evidence

Reply:
What we could do is just save abci.Validator which is only address and power